src/Control/Session.php

```diff
@@ -224,165 +182,69 @@ public function restart(HTTPRequest $request)
-
-            $this->data = isset($_SESSION) ? $_SESSION : array();
-        } else {
-            $this->data = [];
-        }
-
-        // Modify the timeout behaviour so it's the *inactive* time before the session expires.
-        // By default it's the total session lifetime
-        if ($timeout && !headers_sent()) {
-            Cookie::set(session_name(), session_id(), $timeout/86400, $path, $domain ? $domain
-                : null, $secure, true);
-        }
+        $this->getHandler()->start();
+        $this->original = $this->getHandler()->all();
+        return $this;
     }
 
     /**
      * Destroy this session
-     *
-     * @param bool $removeCookie
+     * @return $this
      */
-    public function destroy($removeCookie = true)
+    public function destroy()
     {
-        if (session_id()) {
-            if ($removeCookie) {
-                $path = $this->config()->get('cookie_path') ?: Director::baseURL();
-                $domain = $this->config()->get('cookie_domain');
-                $secure = $this->config()->get('cookie_secure');
-                Cookie::force_expiry(session_name(), $path, $domain, $secure, true);
-            }
-            session_destroy();
-        }
-        // Clean up the superglobal - session_destroy does not do it.
-        // http://nz1.php.net/manual/en/function.session-destroy.php
-        unset($_SESSION);
-        $this->data = null;
+        $this->getHandler()->invalidate();
+        return $this;
     }
 
     /**
      * Set session value
      *
      * @param string $name
-     * @param mixed $val
+     * @param mixed $value
      * @return $this
      */
-    public function set($name, $val)
+    public function set($name, $value)
     {
-        if (!$this->isStarted()) {
-            throw new BadMethodCallException("Session cannot be modified until it's started");
-        }
-        $var = &$this->nestedValueRef($name, $this->data);
-
-        // Mark changed
-        if ($var !== $val) {
-            $var = $val;
-            $this->markChanged($name);
+        $handler = $this->getHandler();
+        $pieces = explode('.', $name);
+        $key = array_shift($pieces);
+
+        // If the name doesn't include any dots, we can just set the value
+        if (empty($pieces)) {
+            $handler->set($name, $value);
+            return $this;
         }
-        return $this;
-    }
 
-    /**
-     * Mark key as changed
-     *
-     * @internal
-     * @param string $name
-     */
-    protected function markChanged($name)
-    {
-        $diffVar = &$this->changedData;
-        foreach (explode('.', $name) as $namePart) {
-            if (!isset($diffVar[$namePart])) {
-                $diffVar[$namePart] = [];
+        // Traverse down existing array (adding placeholders if needed) to set the value
+        $existing = $handler->get($key);
+        $var =& $existing;
+        foreach ($pieces as $namePart) {
+            if (!isset($var) || !is_array($var)) {
```

Comment:
is this a change in behaviour, it feels dangerous as if I do `Session::set('Cart.Size', 3)` and then `Session::set('Card.Size.Total', 7)` then the 'Cart.Size' value is removed and replaced with an array...

I'm not sure how that should work now, but maybe it's a regression?